sips/sip-018/sip-018-signed-structured-data.md

```diff
@@ -0,0 +1,386 @@
+SIP018 message.
+
+![Wallet SIP018 concept](SIP-018-001.png)
+
+# Specification
+
+The challenge lies in producing messages that are both meaningful to humans as
+well as easy to be processed on-chain. Luckily, Clarity is a strongly-typed
+interpreted language. Structured Data is therefore a Clarity Value expression as
+detailed in _SIP002: Smart Contract Language_. These value expressions are
+encoded in Stacks wire format, as detailed in _SIP005: Blocks and Transactions_,
+and then hashed with SHA256. The resulting hash is used as the input for
+signing.
+
+## Definitions
+
+_Clarity Value_: a native Clarity value as expressed in _SIP002: Smart Contract
+Language_.
+
+_Meta Transaction_: a meta transaction can be understood of as a transaction
+that contains another "transaction". For example, a user may sign a message
+that, when received by a particular smart contract, causes it to transfer
+tokens. A transaction carrying the signed message will then trigger the
+underlying action when it is broadcast.
+
+_SHA256_: a cryptographic hash function producing hashes with a length of 32
+bytes.
+
+_Signature_: a proof produced by a signing algorithm, in this case ECDSA with
+the `secp256k1` curve, encoded in RSV order.
+
+_Structured Data_: A structured representation of a message to sign expressed as
+a Clarity Value.
+
+_Wire format_: the underlying encoding for Clarity Values when generating a
+transaction, as expressed in _SIP005: Blocks and Transactions_.
+
+## Formal specification
+
+To obtain a signature _S_ of structured data _D_ using private key _K_:
+
+- `S = signStructuredData(D, K)`; where,
+
+- `signStructuredData(D, K) = sign(messageHash(D), K)`; where `sign` produces a
+  `secp256k1` signature of a message hash using private key _K_; and,
+
+- `messageHash(D) = sha256(structuredDataPrefix || domainHash || structuredDataHash(D))`;
+  where `structuredDataPrefix` is a static value of `"\x53\x49\x50\x30\x31\x38"`
+  (which spells "SIP018" in ASCII), and `domainHash` and `structuredDataHash`
+  are as described below.
+
+**Definition of** `domainHash`
+
+The domain hash ensures that messages are chain and application specific. A
+`domainHash` is generated out of a `domain` Clarity Value:
+
+- `domainHash = structuredDataHash(domain)` where `structuredDataHash` and
+  `domain` are as described below.
+
+**Definition of** `structuredDataHash`
+
+The `structuredDataHash` function calculates a hash _H_ of the input Clarity
+Value _D_.
+
+- `H = structuredDataHash(D)`; where,
+- `structuredDataHash(D) = sha256(ToCVWireFormat(D))`; where `ToCVWireFormat` is
+  a function that encodes an input Clarity Value to its wire format
+  representation as described in _SIP005_.
+
+**Definition of** `domain`
+
+The `domain` is a `tuple` type Clarity Value that represents the domain in which
+messages are signed. Its Clarity type definition is as follows:
+
+```clojure
+{
+	name: (string-ascii len)
+	version: (string-ascii len)
```

Comment:
Could there be disadvantages of excluding utf strings as a data type for `name`? i.e. not allowing Japanese/etc. characters. Internet hostnames get around this by using [Punycode](https://en.wikipedia.org/wiki/Punycode) which is an ascii subset — might be good enough but probably depends on how the signed data is intended to be used (more specifically, how the `domain` is additionally matched).

Maybe not an issue, but I wanted to bring it up...

Reply:
ASCII was chosen on purpose because I assumed that the name in the domain object would be used for "app identity" purposes by the wallet. I wanted to prevent people mimicking different names using UTF8 tricks. I think a future SIP could define a proper "display name" or similar, as you say.

However, we would not want limit the SIP arbitrarily in favour of "Western language". What do you think considering the above?

Reply:
Okay, thanks — seems reasonable and important for user safety.